Add generalized tupled functions abstraction (#6568)
lampepfl/dotty: docs/docs/reference/other-new-features/tupled-function.md

---
layout: doc-page
title: "Tupled Function"
---

Tupled Function
----------------------

With functions bounded to arities up to 22 it was possible to generalize some operation on all function types using overloading. 
Now that we have functions and tuples generalized to [arities above 22](https://dotty.epfl.ch/docs/reference/dropped-features/limit22.html) overloading is not an option anymore. 

> The type class `TupleFunction` provides a way to abstract directly over functions of any arity converting it to an equivalent function that receives all arguments in a single tuple.

Reviewer: ```suggestion
The type class `TupleFunction` provides a way to abstract directly over a function of any arity converting it to an equivalent function that receives all arguments in a single tuple.
```

Author: Updated

```scala
/** Type class relating a `FunctionN[..., R]` with an equvalent tupled function `Function1[TupleN[...], R]`
 *
 *  @tparam F a function type
 *  @tparam G a tupled function type (function of arity 1 receiving a tuple as argument)
 */
@implicitNotFound("${F} cannot be tupled as ${G}")
sealed trait TupledFunction[F, G] {

>   def tuple(f: F): G

Reviewer: `untuple` is missing.

Author: Added

}
```

> The compiler will synthesize an instance of `TupledFunction[F, G]` if:

Reviewer: Should we also indicate that if the type `G` is not instantiated, the compiler will instantiate it so that it is aligned with the type of `F`?

Reviewer: (actually, it seems to work in both directions: if one of `F` or `G` is unknown, the compiler computes it according to the other type)

Author: Yes, it works in both directions. I had to make it infer `F` from `G` to make `untulpe` work

Author: I added it in the documentation

* `F` is a function type of arity `N`
* `G` is a function with a single tuple argument of size `N` and it's types are equal to the arguments of `F`
* The return type of `F` is equal to the return type of `G`

> * `F` and `G` are the same kind of functions (`given` arguments or not)

Reviewer: This line is not clear to me, can you elaborate, please?

Author: If both are normal functions `FunctionN` or both are `ImplicitFunctionN`. How would you stat it?

Reviewer: ```suggestion
* both `F` and `G` have type `FunctionN` or they both have type `ImplicitFunctionN`
```
?

Author: I updated the documentation. Is it clearer now?

Author: Now I used
*  `F` and `G` are the same kind of function (both are `(...) => R` or both are `given (...) => R`)
